benchmarks/src/main/java/org/elasticsearch/benchmark/stream/StreamInputBenchmark.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.benchmark.stream;
+
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.InputStreamStreamInput;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+@Fork(10)
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Benchmark)
+@SuppressWarnings("unused") //invoked by benchmarking framework
+public class StreamInputBenchmark {
```

Comment:
Will drop it. I did however run this benchmark once with the version you reviewed.

Master:

```
On master:
# Run complete. Total time: 01:02:39

Benchmark                                      (size)  Mode  Cnt          Score         Error  Units
StreamInputBenchmark.measureReadAndSumVInt          4  avgt  100         35.233 ±       0.244  ns/op
StreamInputBenchmark.measureReadAndSumVInt         16  avgt  100         85.349 ±       1.447  ns/op
StreamInputBenchmark.measureReadAndSumVInt         64  avgt  100        235.382 ±       2.569  ns/op
StreamInputBenchmark.measureReadAndSumVInt        256  avgt  100        902.952 ±       9.297  ns/op
StreamInputBenchmark.measureReadAndSumVInt       1024  avgt  100       3499.609 ±      38.687  ns/op
StreamInputBenchmark.measureReadAndSumVInt      16348  avgt  100      54520.067 ±     371.619  ns/op
StreamInputBenchmark.measureReadAndSumVInt    1048576  avgt  100    3374447.046 ±   33979.659  ns/op
StreamInputBenchmark.measureReadAndSumVInt   10485760  avgt  100   34004175.888 ±  170539.485  ns/op
StreamInputBenchmark.measureReadAndSumVInt  104857600  avgt  100  344504176.762 ± 4489346.992  ns/op
StreamInputBenchmark.measureReadToArray             4  avgt  100         44.798 ±       0.536  ns/op
StreamInputBenchmark.measureReadToArray            16  avgt  100         96.208 ±       1.272  ns/op
StreamInputBenchmark.measureReadToArray            64  avgt  100        252.820 ±       3.105  ns/op
StreamInputBenchmark.measureReadToArray           256  avgt  100        920.067 ±       6.403  ns/op
StreamInputBenchmark.measureReadToArray          1024  avgt  100       3587.619 ±      31.852  ns/op
StreamInputBenchmark.measureReadToArray         16348  avgt  100      56473.895 ±     386.739  ns/op
StreamInputBenchmark.measureReadToArray       1048576  avgt  100    3475751.564 ±   45444.250  ns/op
StreamInputBenchmark.measureReadToArray      10485760  avgt  100   34633952.391 ±  363944.894  ns/op
StreamInputBenchmark.measureReadToArray     104857600  avgt  100  347333659.817 ± 3212549.852  ns/op
```

This branch:

```
# Run complete. Total time: 01:02:34

Benchmark                                      (size)  Mode  Cnt          Score         Error  Units
StreamInputBenchmark.measureReadAndSumVInt          4  avgt  100         36.267 ±       0.785  ns/op
StreamInputBenchmark.measureReadAndSumVInt         16  avgt  100         85.695 ±       0.678  ns/op
StreamInputBenchmark.measureReadAndSumVInt         64  avgt  100        236.789 ±       1.683  ns/op
StreamInputBenchmark.measureReadAndSumVInt        256  avgt  100        899.875 ±       9.242  ns/op
StreamInputBenchmark.measureReadAndSumVInt       1024  avgt  100       3502.975 ±      31.820  ns/op
StreamInputBenchmark.measureReadAndSumVInt      16348  avgt  100      55528.773 ±     286.482  ns/op
StreamInputBenchmark.measureReadAndSumVInt    1048576  avgt  100    3432045.939 ±   23759.513  ns/op
StreamInputBenchmark.measureReadAndSumVInt   10485760  avgt  100   34495218.291 ±  179885.969  ns/op
StreamInputBenchmark.measureReadAndSumVInt  104857600  avgt  100  349996045.803 ± 1013571.648  ns/op
StreamInputBenchmark.measureReadToArray             4  avgt  100         45.405 ±       0.918  ns/op
StreamInputBenchmark.measureReadToArray            16  avgt  100         96.805 ±       0.988  ns/op
StreamInputBenchmark.measureReadToArray            64  avgt  100        254.719 ±       2.286  ns/op
StreamInputBenchmark.measureReadToArray           256  avgt  100        937.991 ±       7.558  ns/op
StreamInputBenchmark.measureReadToArray          1024  avgt  100       3656.997 ±      29.638  ns/op
StreamInputBenchmark.measureReadToArray         16348  avgt  100      57338.926 ±     292.516  ns/op
StreamInputBenchmark.measureReadToArray       1048576  avgt  100    3497268.663 ±   24068.370  ns/op
StreamInputBenchmark.measureReadToArray      10485760  avgt  100   35388540.405 ±  447633.734  ns/op
StreamInputBenchmark.measureReadToArray     104857600  avgt  100  357208368.590 ± 3177542.818  ns/op
```